fix: invalid parentUser id on OIDC auto expiration

hellivan opened the pull request:
## Description
This commit should fix a bug that was introduced in
`f6d13f57bb1218b60db8808a15ae5485f1ef362b` where the wrong
parentUserId was used for deletion of expired OIDC users

## Motivation and Context
Keycloak IDP user removal did not delete any expired users since `DeleteUser` method on `IAMStoreSys` was
called with Keycloak user ids instead of Minio's internal user id.

## How to test this PR?
You could setup a Minio deployment with Keycloak as its IDP. After that some service accounts should be created in Minio using a Keycloak user. If this user gets disabled or deleted in Keycloak, the services accounts still exists in Minio and can be used to access the system.
For now I did not implement any unit tests since I'm not 100% sure if implementation satisfies your general implementation/coding strategy in Minio. Please let me know if implementation approach is ok for you
I can implement the missing unit tests.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [X] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

harshavardhana commented on cmd/iam-store.go:
```diff
@@ -1448,33 +1448,35 @@ func (store *IAMStoreSys) DeleteUsers(ctx context.Context, users []string) error
-				}
-			}
-			res.Add(parentUser)
+			parentUserSessionTokenMap[cred.ParentUser] = cred.SessionToken
```
This looks completely wrong cred.ParentUser is not meaningful in any manner here - you are returning way too many entries here to expire only a sub-set of entries. 

The previous code actually correctly handles this.

hellivan replied:
In theory the method returns the same set of entries as before (since we are using a Map). The change here is that the method returns a `ParentUserInfo` struct instead of plain userid, which is required in order to extract OIDC user id when fetching user information at some later point (`LookupUser`).

harshavardhana reviewed (changes requested):
I see how this should be implemented you are making way too many changes here to fix a small bug. 

Its actually not a regression since the previous fix was to fix another bug that never deleted the service accounts anyways, so that was kind of broken for a while.

harshavardhana reviewed (changes requested):
All you need is this 

```diff
diff --git a/cmd/iam-store.go b/cmd/iam-store.go
index 63b94d018..1e719afca 100644
--- a/cmd/iam-store.go
+++ b/cmd/iam-store.go
@@ -1450,11 +1450,11 @@ func (store *IAMStoreSys) DeleteUsers(ctx context.Context, users []string) error
 
 // GetAllParentUsers - returns all distinct "parent-users" associated with STS or service
 // credentials.
-func (store *IAMStoreSys) GetAllParentUsers() []string {
+func (store *IAMStoreSys) GetAllParentUsers() map[string]string {
        cache := store.rlock()
        defer store.runlock()
 
-       res := set.NewStringSet()
+       res := map[string]string{}
        for _, cred := range cache.iamUsersMap {
                if cred.IsServiceAccount() || cred.IsTemp() {
                        parentUser := cred.ParentUser
@@ -1470,11 +1470,13 @@ func (store *IAMStoreSys) GetAllParentUsers() []string {
                                        }
                                }
                        }
-                       res.Add(parentUser)
+                       if _, ok := res[parentUser]; !ok {
+                               res[parentUser] = cred.ParentUser
+                       }
                }
        }
 
-       return res.ToSlice()
+       return res
 }
 
 // SetUserStatus - sets current user status.
diff --git a/cmd/iam.go b/cmd/iam.go
index 4669a505a..7371b20dc 100644
--- a/cmd/iam.go
+++ b/cmd/iam.go
@@ -1003,7 +1003,7 @@ func (sys *IAMSys) SetUserSecretKey(ctx context.Context, accessKey string, secre
 func (sys *IAMSys) purgeExpiredCredentialsForExternalSSO(ctx context.Context) {
        parentUsers := sys.store.GetAllParentUsers()
        var expiredUsers []string
-       for _, parentUser := range parentUsers {
+       for parentUser, expiredUser := range parentUsers {
                u, err := globalOpenIDConfig.LookupUser(parentUser)
                if err != nil {
                        logger.LogIf(GlobalContext, err)
@@ -1012,7 +1012,7 @@ func (sys *IAMSys) purgeExpiredCredentialsForExternalSSO(ctx context.Context) {
                // If user is set to "disabled", we will remove them
                // subsequently.
                if !u.Enabled {
-                       expiredUsers = append(expiredUsers, parentUser)
+                       expiredUsers = append(expiredUsers, expiredUser)
                }
        }
 
@@ -1025,12 +1025,12 @@ func (sys *IAMSys) purgeExpiredCredentialsForExternalSSO(ctx context.Context) {
 func (sys *IAMSys) purgeExpiredCredentialsForLDAP(ctx context.Context) {
        parentUsers := sys.store.GetAllParentUsers()
        var allDistNames []string
-       for _, parentUser := range parentUsers {
+       for parentUser, expiredUser := range parentUsers {
                if !globalLDAPConfig.IsLDAPUserDN(parentUser) {
                        continue
                }
 
-               allDistNames = append(allDistNames, parentUser)
+               allDistNames = append(allDistNames, expiredUser)
        }
 
        expiredUsers, err := globalLDAPConfig.GetNonEligibleUserDistNames(allDistNames)
```

hellivan committed:
fix: invalid parentUser id on OIDC auto expiration

This commit should fix a bug that was introduced in
`f6d13f57bb1218b60db8808a15ae5485f1ef362b` where the wrong
parentUserId was used for deletion of expired OIDC users
